[JIT] dict type unification fix

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32185 [JIT] dict type unification fix**

Previously we would unify the contained types of dictionaries, however this breaks type safety.
```
@torch.jit.script
def test(input: Dict[str, None], cond):
    if cond:
        out = input
    else:
        out: {"1": 1}
    out["hi"] = 3
```
   
This would only occur if a dictionary is being re-assigned across an if condition with different contained types, which is pretty unlikely. I tested `model_backward_compatibility` for all fb models and this didn't break anything. This PR is a precursor to alias analysis changes. 

Also fixes `Future` type unification. Because `Future` is an immutable type, it is okay to unify the contained type.

Differential Revision: [D19398585](https://our.internmc.facebook.com/intern/diff/D19398585)

Commit:
type unification fixes

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit be036d17:


* **2/2** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=be036d176f246e18935320ba941ccf011c775260).

### :detective: 2 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4243076) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_simple_test (1/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86092886) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86092886))
<details>
<summary>
<code>Jan 14 20:59:16 RuntimeError: test_jit_simple failed!</code>
</summary>

```
Jan 14 20:59:16 Ran 2326 tests in 81.117s 
Jan 14 20:59:16  
Jan 14 20:59:16 FAILED (errors=1, skipped=79, expected failures=1) 
Jan 14 20:59:16  
Jan 14 20:59:16 Generating XML reports... 
Jan 14 20:59:16 Traceback (most recent call last): 
Jan 14 20:59:16   File "test/run_test.py", line 456, in <module> 
Jan 14 20:59:16     main() 
Jan 14 20:59:16   File "test/run_test.py", line 449, in main 
Jan 14 20:59:16     raise RuntimeError(message) 
Jan 14 20:59:16 RuntimeError: test_jit_simple failed! 
Jan 14 20:59:16 + cleanup 
Jan 14 20:59:16 + retcode=1 
Jan 14 20:59:16 + set +x 
Jan 14 20:59:16 =================== sccache compilation log =================== 
Jan 14 20:59:16 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 14 20:59:16 Compile requests                 2 
Jan 14 20:59:16 Compile requests executed        0 
Jan 14 20:59:16 Cache hits                       0 
Jan 14 20:59:16 Cache misses                     0 
Jan 14 20:59:16 Cache timeouts                   0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4243072) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_legacy_test (2/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86093531) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86093531))
<details>
<summary>
<code>Jan 14 21:00:07 RuntimeError: test_jit_legacy failed!</code>
</summary>

```
Jan 14 21:00:07 Ran 2326 tests in 129.324s 
Jan 14 21:00:07  
Jan 14 21:00:07 FAILED (errors=1, skipped=62, expected failures=1) 
Jan 14 21:00:07  
Jan 14 21:00:07 Generating XML reports... 
Jan 14 21:00:07 Traceback (most recent call last): 
Jan 14 21:00:07   File "test/run_test.py", line 456, in <module> 
Jan 14 21:00:07     main() 
Jan 14 21:00:07   File "test/run_test.py", line 449, in main 
Jan 14 21:00:07     raise RuntimeError(message) 
Jan 14 21:00:07 RuntimeError: test_jit_legacy failed! 
Jan 14 21:00:07 + cleanup 
Jan 14 21:00:07 + retcode=1 
Jan 14 21:00:07 + set +x 
Jan 14 21:00:07 =================== sccache compilation log =================== 
Jan 14 21:00:08 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 14 21:00:08 Compile requests               250 
Jan 14 21:00:08 Compile requests executed        0 
Jan 14 21:00:08 Cache hits                       0 
Jan 14 21:00:08 Cache misses                     0 
Jan 14 21:00:08 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "[JIT] dict type unification fix"


Previously we would unify the contained types of dictionaries, however this breaks type safety.
```
@torch.jit.script
def test(input: Dict[str, None], cond):
    if cond:
        out = input
    else:
        out: {"1": 1}
    out["hi"] = 3
```
   
This would only occur if a dictionary is being re-assigned across an if condition with different contained types, which is pretty unlikely. I tested `model_backward_compatibility` for all fb models and this didn't break anything. This PR is a precursor to alias analysis changes. 

Also fixes `Future` type unification. Because `Future` is an immutable type, it is okay to unify the contained type. 

    



[ghstack-poisoned]

suo commented on aten/src/ATen/core/type.cpp:
```diff
@@ -202,9 +202,17 @@ c10::optional<TypePtr> unifyTypes(const TypePtr& t1, const TypePtr& t2) {
     return static_cast<TypePtr>(TupleType::create(elements));
   }
 
+  if (t1->cast<FutureType>() && t2->cast<FutureType>()) {
```
perhaps I'm missing something: why are Futures exempt from the same problem that we had with dicts? e.g. isn't it wrong for `Future[int]` and `Future[Optional[int]]` to unify into `Future[int]`?

eellison replied:
Future is an immutable type, there's no way to mutate what it's contained element is.

suo replied:
ah, ok I buy it.

eellison replied:
However I tried an example and i got an assert failure from elsewhere, so it looks like we currently enforce that Future is waited on in the block it is created:
```
def int_fn():
    return 1

def none_fn():
    return None

@torch.jit.script
def test(cond):
    if cond:
        x = torch.jit._fork(int_fn)
    else:
        x = torch.jit._fork(none_fn)
    return torch.jit._wait(x)
```

> RuntimeError: fork_graph_context->uses().size() == 1 INTERNAL ASSERT FAILED at ../torch/csrc/jit/passes/inline_forked_closures.cpp:43, please report a bug to PyTorch.  (inlineForkedClosure at ../tor

So i can remove this if requested.
